plugins/out_azure_kusto/azure_kusto_conf.c

```diff
@@ -205,136 +194,139 @@ static int parse_storage_resources(struct flb_azure_kusto *ctx, struct flb_confi
 
     jsmn_init(&parser);
     tokens = flb_calloc(1, sizeof(jsmntok_t) * tok_size);
-    if (!tokens) {
-        flb_plg_error(ctx->ins, "error allocating tokens");
-        flb_sds_destroy(resource_uri);
-        return -1;
-    }
-
-    ret = jsmn_parse(&parser, response, flb_sds_len(response), tokens, tok_size);
-    if (ret <= 0) {
-        flb_plg_error(ctx->ins, "error parsing JSON response: %s", response);
-        flb_sds_destroy(resource_uri);
-        flb_free(tokens);
-        return -1;
-    }
 
-    /* skip all tokens until we reach "Rows" */
-    for (i = 0; i < ret - 1; i++) {
-        t = &tokens[i];
-
-        if (t->type != JSMN_STRING) {
-            continue;
-        }
-
-        token_str = response + t->start;
-        token_str_len = (t->end - t->start);
-
-        /**
-         * if we found the Rows key, skipping this token and the next one (key and
-         * wrapping array value)
-         */
-        if (token_str_len == 4 && strncmp(token_str, "Rows", 4) == 0) {
-            i += 2;
-            break;
-        }
-    }
-
-    /* iterating rows, each row will have 3 tokens: the array holding the column values,
-     * the first value containing the resource type, and the second value containing the
-     * resource uri */
-    for (; i < ret; i++) {
-        t = &tokens[i];
-
-        /**
-         * each token should be an array with 2 strings:
-         * First will be the resource type (TempStorage, SecuredReadyForAggregationQueue,
-         * etc...) Second will be the SAS URI
-         */
-        if (t->type != JSMN_ARRAY) {
-            break;
-        }
-
-        /* move to the next token, first item in the array - resource type */
-        i++;
-        t = &tokens[i];
-        if (t->type != JSMN_STRING) {
-            break;
-        }
-
-        token_str = response + t->start;
-        token_str_len = (t->end - t->start);
-
-        flb_plg_debug(ctx->ins, "found resource of type: %.*s ", t->end - t->start,
-                      response + t->start);
-
-        if (token_str_len == 11 && strncmp(token_str, "TempStorage", 11) == 0) {
-            resource_type = AZURE_KUSTO_RESOURCE_STORAGE;
+    if (tokens) {
+        ret = jsmn_parse(&parser, response, flb_sds_len(response), tokens, tok_size);
+
+        if (ret > 0) {
+            /* skip all tokens until we reach "Rows" */
+            for (i = 0; i < ret - 1; i++) {
+                t = &tokens[i];
+
+                if (t->type != JSMN_STRING) {
+                    continue;
+                }
+
+                token_str = response + t->start;
+                token_str_len = (t->end - t->start);
+
+                /**
+                 * if we found the Rows key, skipping this token and the next one (key and
+                 * wrapping array value)
+                 */
+                if (token_str_len == 4 && strncmp(token_str, "Rows", 4) == 0) {
```

Comment:
Is this a list of pairs? if that's the case, wouldn't we want to set the stepping for `i` to 2 in the `for` statement so we don't compare the value elements to avoid false positives and to minimize how many comparisons we perform?

Reply:
I was talking about the for loop that spans from 203 to 221 but it's not important, Do I properly understand that this is stream processing a JSON message looking for a specific array stored in a dict key named "Rows"?

I don't think this is a problem since it's one of those calculated risks taken based on data format knowledge which coming from someone intimately related to the system is acceptable.

I was just thinking about the performance hit of this type of thing and json overall and wondering if something loosely related to an xpath accessor system would be a good choice to optimally get to the desired data in an unknown json buffer using a streaming decoder. I hope my ramble is clear enough, json support / performance is something I'm interested in improving in te long run.

Reply:
> Do I properly understand that this is stream processing a JSON message looking for a specific array stored in a dict key named "Rows"?

Yes, but the tokens are stored flat, so there's no guarantee for a key-value pairs (if a value is an array or object, it will be followed by it's sub-tokens, followed by the next key, and so forth).

Reply:
Yes, that's what I noticed in the code, from a practical standpoint it's quite similar to how msgpack is handled in stream mode by mpack. I wrote an abstraction layer for it in cmetrics, it has a different aim but I can see how this would work. 

Interesting topic.